compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -198,6 +200,55 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let pick =
             self.lookup_probe(span, segment.ident, self_ty, call_expr, ProbeScope::TraitsInScope)?;
 
+        if span.edition() < Edition::Edition2021 {
+            if let sym::try_into = segment.ident.name {
+                if let probe::PickKind::TraitPick = pick.kind {
+                    if !matches!(self.tcx.crate_name(pick.item.def_id.krate), sym::std | sym::core)
+                    {
+                        self.tcx.struct_span_lint_hir(
+                            FUTURE_PRELUDE_COLLISION,
+                            call_expr.hir_id,
+                            call_expr.span,
+                            |lint| {
+                                let sp = call_expr.span;
+                                let trait_name =
+                                    self.tcx.def_path_str(pick.item.container.assert_trait());
+
+                                let mut lint = lint.build(&format!(
+                                    "trait method `{}` will become ambiguous in Rust 2021",
+                                    segment.ident.name
+                                ));
+
+                                if let Ok(self_expr) =
+                                    self.sess().source_map().span_to_snippet(self_expr.span)
+                                {
+                                    lint.span_suggestion(
+                                        sp,
+                                        "disambiguate the associated function",
+                                        format!(
+                                            "{}::{}({})",
+                                            trait_name, segment.ident.name, self_expr,
```

Comment:
Reading this again, more carefully this time, it seems clear that *just* including `self_expr` is not sufficient, as there may be more arguments. 

Similarly, we will want to add the auto-ref and other adjustments for `self_expr`. Those can be found in the [autoderefs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/method/probe/struct.Pick.html#structfield.autoderefs) and [autoref_or_ptr_adjustment](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/method/probe/struct.Pick.html#structfield.autoref_or_ptr_adjustment) fields. Basically if the autoref is `Some` we would add a `&` (or `raw_addr!`) and then N `*` where `N` is the value of the autoderefs field.

Reply:
I had just added the autoderefs before seeing this, but didn't account for `autoref_or_ptr_adjustment`, good catch, thanks!